MediaLibrary: Refactoring web search API (Lombiq Technologies: ORCH-206), fixes #7118

I3undy opened the pull request.

I3undy committed:
Refactoring out Orchard.MediaLibrary WebSearch into separated module and implementing pixabay's API for image search instead of Bing

I3undy committed:
Adding Bing image search support too

I3undy committed:
Starting the modularization of the WebSearch views and JS

I3undy committed:
Merge branch 'dev' into issue/ORCH-206

I3undy committed:
Moving Media WebSearch Api calls to the server-side, we want to hide the API Keys

I3undy committed:
Cleaning up websearch model properties and making the websearch-properties-selection great again

I3undy committed:
Moving Web search API data bindings to the server-side, the api returns with uniformed Json result

I3undy committed:
Adding IWebSearchSettings for the purpose of beautiful code

Also adding disabled option for the providers without ApiKey

I3undy committed:
Code styling

I3undy committed:
Cleaning up Web.config

I3undy committed:
Using IEnumerable as return type for WebSearchProvider and moving Settings into models namespace

Also removing unnecessary CSS files

I3undy committed:
Adding null checks, code styling, filling Module.txt, removing unnecessary folder, adding CSS styling

I3undy committed:
Code styling

I3undy committed:
Important code styling

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/signed)](https://cla.dotnetfoundation.org/OrchardCMS/Orchard?pullRequest=8229) <br/>All CLA requirements met.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Controllers/Api/WebSearchApiController.cs:
```diff
@@ -0,0 +1,32 @@
+
+        [HttpGet]
+        public IHttpActionResult GetImages(string query, string providerType) {
+            if (!_authorizer.Authorize(Permissions.ManageMediaContent)) {
```
Let's use a separate, custom permission just for the Web Search instead of ManageMediaContent.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Module.txt:
```diff
@@ -0,0 +1,23 @@
+Website: https://orchardproject.net
+Version: 1.0
+OrchardVersion: 1.10.3
+Description: Enables Orchard to use image search in media or as a provider.
```
This line can be removed.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Module.txt:
```diff
@@ -0,0 +1,23 @@
+Description: Enables Orchard to use image search in media or as a provider.
+Features:
+    Orchard.MediaLibrary.WebSearch:
+        Description: Enables Orchard to use image search in media or as a provider.
```
Base functionality for Orchard to be able import images from web search results to the Media Library.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Module.txt:
```diff
@@ -0,0 +1,23 @@
+        Name: Media Library Web Search
+        Dependencies: Orchard.MediaLibrary
+    Orchard.MediaLibrary.WebSearch.Bing:
+        Description: Enables Orchard to use Bing image search in media or as a provider.
```
Bing web search provider for importing images to the Media Library.
(similarly for Pixabay too)

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Providers/BingWebSearchProvider.cs:
```diff
@@ -0,0 +1,55 @@
+            var webSearchResult = new List<WebSearchResult>();
+
+            foreach (var hit in apiResult.value) {
+                String imageSize = hit.contentSize;
```
"string" or var ... as string.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Providers/IWebSearchProvider.cs:
```diff
@@ -0,0 +1,13 @@
+        string Name { get; }
+        IWebSearchSettings Settings { get; }
+        bool IsValid { get; }
+        string ApiKey { get; }
```
This is already exposed through the Settings property, but a shortcut can be added as an extension method.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Providers/IWebSearchProvider.cs:
```diff
@@ -0,0 +1,13 @@
+    public interface IWebSearchProvider : IDependency {
+        string Name { get; }
+        IWebSearchSettings Settings { get; }
+        bool IsValid { get; }
```
This is unnecessary for the interface, should be an extension method instead, since the ApiKey is exposed through the Settings property.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Views/EditorTemplates/Parts/WebSearch.BingWebSearchSettings.cshtml:
```diff
@@ -0,0 +1,11 @@
+        <span class="hint">@T("Your private key.")</span>
+    </div>
+
+    @T("Get a free API Key on {0}", Html.Raw("<a href=\"https://azure.microsoft.com/hu-hu/services/cognitive-services/bing-web-search-api\">https://azure.microsoft.com/hu-hu/services/cognitive-services/bing-web-search-api/</a>"))
```
This is the Hungarian URL - I'm not sure if it redirects according to the browser culture, but we should use en-US as default.

BenedekFarkas replied:
Link text should be just "Azure".

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Views/EditorTemplates/Parts/WebSearch.PixabayWebSearchSettings.cshtml:
```diff
@@ -0,0 +1,11 @@
+        <span class="hint">@T("Your private key.")</span>
+    </div>
+
+    @T("Get a free API Key on {0}", Html.Raw("<a href=\"https://pixabay.com/api/docs\">https://pixabay.com/api/docs</a>"))
```
Link text should be just "Pixabay".

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Views/WebSearch/Index.cshtml:
```diff
@@ -1,90 +1,102 @@
+        @if (!hasValidApiKey)
+        {
             <div id="message-api">
                 @T("You need to define an API Key in the Media settings before being able to use this feature.")
```
Please update this according to new functionality.

I3undy replied:
I'm not sure which new functionality are you referring to. I made another warning message if there are no Web Search providers enabled.

BenedekFarkas replied:
Yep, that's what I was referring to!

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Module.txt:
```diff
@@ -0,0 +1,23 @@
+﻿Name: Orchard.MediaLibrary.WebSearch
+AntiForgery: enabled
+Author: The Orchard Team
```
The Orchard Team, Lombiq Technologies Ltd.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/IBingApi.cs:
```diff
@@ -0,0 +1,10 @@
+        [Get("bing/v7.0/images/search")]
+        Task<string> GetImagesAsync([Header("Ocp-Apim-Subscription-Key")] string apiKey, string q);
+    }
+}
```
Let's move this to the Api folder.

I3undy replied:
I'm not sure which Api folder.

BenedekFarkas replied:
In a new folder in the module, but maybe Services would be better.

I3undy replied:
Okay, It's there.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/IPixabayApi.cs:
```diff
@@ -0,0 +1,10 @@
+        [Get("api/")]
+        Task<string> GetImagesAsync(string key, string q);
+    }
+}
```
Let's move this to the Api folder.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Module.txt:
```diff
@@ -0,0 +1,23 @@
+    Orchard.MediaLibrary.WebSearch.Bing:
+        Description: Enables Orchard to use Bing image search in media or as a provider.
+        Category: Media
+        Name: Bing
```
Media Library Web Search - Bing

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Module.txt:
```diff
@@ -0,0 +1,23 @@
+    Orchard.MediaLibrary.WebSearch.Pixabay:
+        Description: Enables Orchard to use Pixabay image search in media or as a provider.
+        Category: Media
+        Name: Pixabay
```
Media Library Web Search - Pixabay

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.MediaLibrary.WebSearch/Views/EditorTemplates/Parts/WebSearch.BingWebSearchSettings.cshtml:
```diff
@@ -0,0 +1,11 @@
+<fieldset>
+    <legend>@T("Bing Web Search")</legend>
+    <div>
+        @Html.TextBoxFor(m => m.ApiKey, new { @class = "text medium"})
```
This should be handled as a password, like the SMTP-settings (same for Pixabay).

BenedekFarkas reviewed (changes requested):
~~menu.websearch-mediaproviders.css should be moved from MediaLibrary to the new module and included when necessary.~~

BenedekFarkas reviewed (changes requested):
~~When WebSearch is enabled, but there are no providers enabled, the message "You need to define an API Key in the Media settings before being able to use this feature." is shown when clicking on Web Search in the left-side navigation. The message should instead indicate that there are no Web Search providers enabled.
The existing message however is still relevant when the selected provider is invalid.~~

BenedekFarkas reviewed (changes requested):
~~Different kind of UI interactions trigger the following client-side error:~~

> Uncaught TypeError: c.querySelectorAll is not a function at MutationObserver.<anonymous> (content_script_bundle.js:16)

BenedekFarkas reviewed (changes requested):
~~Importing images triggers this client-side error:~~

> GET [...]/Modules/Orchard.MediaLibrary.WebSearch/Styles/images/ajax-loader.gif 404 (Not Found)

I3undy commented:
> Different kind of UI interactions trigger the following client-side error:
> 
> > Uncaught TypeError: c.querySelectorAll is not a function at MutationObserver. (content_script_bundle.js:16)

I can't repro this, but Google says that, it is a [Ghostery bug](https://github.com/webpack/webpack/issues/9335).

I3undy committed:
Renaming the features of MediaLibrary WebSearch

I3undy committed:
Moving WebSearch APIs into the related folders

Moving ajax-loader and menu.websearch-mediaproviders.css to the related module
Adding custom permission for the web search

I3undy committed:
Making the private key of Web Search site settings password to textbox field

I3undy committed:
Adding support for Google image web search

I3undy committed:
Adding Extension method for IsValid and ApiKey of WebSearchProviders

Renaming permission from ManageWebSearchMediaContent to AccessMediaWebSearch
Adding warning if  there are no Web Search providers enabled.

I3undy committed:
Removing random characters, I can't even

I3undy committed:
Suggesting Bing for default WebSearch feature instead of Google

I3undy committed:
Removing WebSearchProviderBase because it has became unnecessary
